### PR TITLE
Add the ability to upload .docx files to the NJ AI Assistant

### PR DESCRIPTION
**Pull request by sannidhishukla**
# Context
With the changes in this branch, users of the NJ AI assistant will now be able to upload .docx files when using the AI assistant.

This branch also includes some minor improvements to the repository:
1. Basic unit tests around uploads as well as new scripts for running these tests separately for the frontend and backend
2. Updates to the NJ README to give some information on running the code + tests locally

## Future Additions
Noting that in a future branch we should continue to add more unit tests + add the ability to run tests through github actions

### CODE REVIEW

File: frontend/src/components/QuestionInput/QuestionInput.test.tsx [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    expect(await axe(container)).toHaveNoViolations();
+  })
+
+  it.each([
+    ['jpg', ACCEPTED_FILE_TYPES.JPEG],
+    ['png', ACCEPTED_FILE_TYPES.PNG],
+    ['gif', ACCEPTED_FILE_TYPES.GIF],
+    ['bmp', ACCEPTED_FILE_TYPES.BMP],
+    ['tiff', ACCEPTED_FILE_TYPES.TIFF]
+  ])(
```
Comment (mluedke2) [inline frontend/src/components/QuestionInput/QuestionInput.test.tsx]: [dust] - i like the `it.each()` and i wonder if the other file types should be put in here too, since it looks like the contents of the PDF, CSV, and DOCX are also the same. So it could be one test for all file types?

File: frontend/src/components/QuestionInput/QuestionInput.tsx [hunk 2]
```diff
@@ -92,25 +92,41 @@ export const QuestionInput = ({ onSend, disabled, placeholder, clearOnSend, conv
...
             extension: selectedFile.type,
             size: selectedFile.size
           })
         }
 
         reader.readAsText(selectedFile)
+      } else if (selectedFile.type === ACCEPTED_FILE_TYPES.DOCX) {
+        try {
+          const arrayBuffer = await selectedFile.arrayBuffer()
+          const extractedText = (await mammoth.extractRawText({ arrayBuffer })).value
```
Comment (mluedke2) [inline frontend/src/components/QuestionInput/QuestionInput.tsx]: [magic dust] - oh that's how one does that!

File: frontend/src/components/QuestionInput/QuestionInput.tsx [hunk 3]
```diff
@@ -92,25 +92,41 @@ export const QuestionInput = ({ onSend, disabled, placeholder, clearOnSend, conv
...
+          }
+        } catch (err) {
+          setInputError('Failed to upload .docx file. Please try uploading a different file.')
+        }
       } else {
         const reader = new FileReader()
         reader.onloadend = () => {
           send({
             name: selectedFile.name,
-            type: FileType.Image,
```
Comment (mluedke2) [inline frontend/src/components/QuestionInput/QuestionInput.tsx:113]: [dust] - why is this line removed?

update: i see that it's removed in the function itself... i guess i transfer my question to there. Is it just no longer needed?

File: frontend/src/pages/chat/Chat.tsx [hunk 4]
```diff
@@ -816,7 +816,7 @@ const Chat = () => {
                       <div className={styles.chatMessageUser} tabIndex={0}>
                         <div className={styles.chatMessageUserMessage}>
                           {answer.uploaded_file != null &&
-                            answer.uploaded_file.type === FileType.Image &&
```
Comment (mluedke2) [inline frontend/src/pages/chat/Chat.tsx:819]: [dust] margin change might get picked up by a code formatter, doesn't really matter otherwise though

**Review by mluedke2 — APPROVED** [review]
cool! a few not-important dusts

File: frontend/src/components/QuestionInput/QuestionInput.tsx (hunk 2 above)
Comment (sannidhishukla) [reply frontend/src/components/QuestionInput/QuestionInput.tsx]: i am surprised to learn how many different packages are needed to extract text/contents from different filetypes!

File: frontend/src/components/QuestionInput/QuestionInput.tsx (hunk 3 above)
Comment (sannidhishukla) [reply frontend/src/components/QuestionInput/QuestionInput.tsx:113]: yeah, i removed `type` as a field within the `UploadedFile` interface. previously it was basically just being used to distinguish between images + non-images but now the `isImageFile()` function handles that.

File: frontend/src/pages/chat/Chat.tsx (hunk 4 above)
Comment (sannidhishukla) [reply frontend/src/pages/chat/Chat.tsx:819]: ooh yes thanks for catching!

File: frontend/src/components/QuestionInput/QuestionInput.test.tsx (hunk 1 above)
Comment (sannidhishukla) [reply frontend/src/components/QuestionInput/QuestionInput.test.tsx]: yes good point! updating now...

File: frontend/src/custom/fileUploadUtils.ts [hunk 5]
```diff
@@ -1,23 +1,21 @@
...
-]
+export enum ACCEPTED_FILE_TYPES {
+  JPEG = 'image/jpeg',
+  PNG = 'image/png',
+  GIF = 'image/gif',
+  BMP = 'image/bmp',
+  TIFF = 'image/tiff',
+  PDF = 'application/pdf',
+  CSV = 'text/csv',
+  DOCX = 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
```
Comment (namanaman) [inline frontend/src/custom/fileUploadUtils.ts:9]: That's wild

File: frontend/src/custom/fileUploadUtils.ts
Comment (namanaman) [inline frontend/src/custom/fileUploadUtils.ts:1]: Nice refactor

File: NJ_README.md [hunk 6]
```diff
@@ -12,3 +12,14 @@ The `nj-stable` branch is deployed to our actual chat application, and we should
...
 5. Run `git push -f` to force push this rebase onto the remote repo. There should not be any commits as part of this push.
 6. You should see that the `nj-stable` branch is no longer behind the `main` branch, and only ahead by the commits the NJ team has made.
+
+## How to run locally
+1. Clone the repo and check out the `nj-stable` branch or any branch created from it.
+2. In the root directory of the repo, create a new `.env` file. Update this file with the values found in Bitwarden.
+3. From the root directory of the repo, run the ./start.sh command
+
+## How to run unit tests locally
+Navigate into the `frontend directory`.
```
Comment (namanaman) [inline NJ_README.md]: slight edit to "Navigate into the `frontend` directory. `cd frontend`"

File: NJ_README.md [hunk 7]
```diff
@@ -12,3 +12,14 @@ The `nj-stable` branch is deployed to our actual chat application, and we should
 4. Run `git pull --rebase origin main` to rebase our NJ-specific changes on top of the latest upstream work. Fix merge conflicts as needed.
 5. Run `git push -f` to force push this rebase onto the remote repo. There should not be any commits as part of this push.
 6. You should see that the `nj-stable` branch is no longer behind the `main` branch, and only ahead by the commits the NJ team has made.
+
+## How to run locally
```
Comment (namanaman) [inline NJ_README.md:16]: Thanks for adding this!

File: frontend/package.json [hunk 8]
```diff
@@ -19,16 +20,21 @@
...
     "lodash-es": "^4.17.21",
+    "mammoth": "^1.8.0",
     "plotly.js": "^2.32.0",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
     "react-markdown": "^7.0.1",
     "react-pdftotext": "^1.3.0",
     "react-plotly.js": "^2.6.0",
     "react-router-dom": "^6.8.1",
+    "react-scripts": "^5.0.1",
```
Comment (namanaman) [inline frontend/package.json]: [Dust] Curious why we use react scripts here? Could we run the tests by directly installing those needed packages? `react-scripts` isn't maintained, so seeing if there's an alternative

File: frontend/package.json [hunk 9]
```diff
@@ -19,16 +20,21 @@
     "@fluentui/react-hooks": "^8.6.29",
     "@fluentui/react-icons": "^2.0.195",
     "@newjersey/feedback-widget": "^0.6.0",
+    "@testing-library/user-event": "^14.5.2",
     "dompurify": "^3.0.8",
+    "identity-obj-proxy": "^3.0.0",
```
Comment (namanaman) [inline frontend/package.json:25]: [Dust] What's this package for?

File: frontend/src/components/QuestionInput/QuestionInput.tsx [hunk 10]
```diff
@@ -2,11 +2,12 @@ import { useRef, useState } from 'react'
 import { Stack, TextField } from '@fluentui/react'
 import { SendRegular } from '@fluentui/react-icons'
 import pdfToText from 'react-pdftotext'
+import mammoth from 'mammoth'
```
Comment (namanaman) [inline frontend/src/components/QuestionInput/QuestionInput.tsx]: [Dust] Can you import `extractRawText` from here instead? or does it not support that destructuring

File: frontend/src/components/QuestionInput/QuestionInput.tsx [hunk 11]
```diff
@@ -136,10 +152,11 @@ export const QuestionInput = ({ onSend, disabled, placeholder, clearOnSend, conv
 
   const onFileChange = (event: React.ChangeEvent<HTMLInputElement>) => {
     const file = event.target.files?.[0]
+
     if (file) {
-      if (!ACCEPTED_FILE_TYPES.includes(file.type)) {
+      if (!(Object.values(ACCEPTED_FILE_TYPES) as string[]).includes(file.type)) {
         setInputError(
-          'Only the following file types are supported: .csv, .pdf, .jpeg, .png, .gif, .bmp, .tiff. Please try a different file.'
+          'Only the following file types are supported: .csv, .doc, .docx, .pdf, .jpeg, .png, .gif, .bmp, .tiff. Please try a different file.'
```
Comment (namanaman) [inline frontend/src/components/QuestionInput/QuestionInput.tsx]: [Pebble] Remove `doc`

File: frontend/src/api/api.test.ts
Comment (namanaman) [inline frontend/src/api/api.test.ts:1]: This is great

File: frontend/src/components/QuestionInput/QuestionInput.test.tsx
Comment (namanaman) [inline frontend/src/components/QuestionInput/QuestionInput.test.tsx:1]: This is great

**Review by namanaman — APPROVED** [review]
Nice work, thanks for the test adds and refactors! A few small things below

File: frontend/package.json (hunk 9 above)
Comment (sannidhishukla) [reply frontend/package.json:25]: this is a transformer for CSS files for when running tests. there's a built in version in create react app but unfortunately it needs to be pulled in separately here.

File: frontend/src/components/QuestionInput/QuestionInput.tsx (hunk 10 above)
Comment (sannidhishukla) [reply frontend/src/components/QuestionInput/QuestionInput.tsx]: yes! that works!

File: frontend/package.json (hunk 9 above)
Comment (sannidhishukla) [reply frontend/package.json:25]: it's used in `jest.config.ts`

**Comment by ruthienachmany** [comment]
is this in prod yet?
